### ISSUE TITLE
Cocoa ns-image class doesn't implement image>stream

### ISSUE THREAD

**Issue by phred** [factor/factor#2355]
Both the GDI+ and GTK backends do:
https://github.com/factor/factor/blob/master/basis/images/loader/gtk/gtk.factor#L95-L96
https://github.com/factor/factor/blob/master/basis/images/loader/gdiplus/gdiplus.factor#L109-L110

However this is missing from `images.loader.cocoa` backend so `save-graphics-image` for example won't work. Poking around a little it does seem that the necessary pieces are implemented to write images, it's just the `image>stream` implementation that's missing:

```factor
USE: core-foundation.launch-services
! (lots of other USE omitted, the above vocab provides the kUTTypePNG constant)

sans-serif "hello there"
string>image
drop
"/tmp/output.png"
kUTTypePNG
save-ns-image
```
I think all that's needed here is a word to map from the internal image classes to the kUTType* CoreFoundation constants. Looking at the other backends this should be a pretty small implementation. I'm happy to tackle this, as I want the ability to write images for a thing I'm working on!

**Comment by mrjbq7** [comment]
That would be a great contribution!

**Comment by AlexIljin** [comment]
It's interesting, but the example below works fine for me in Windows (although for some reason the image is flipped top to bottom).
```
USING: fonts images.loader ui.text ;
monospace-font "Hello, there!" string>image drop 
"./hello.jpg" save-graphic-image
```
But if I change the file extension to ".png", then the saved image is completely white. Should I create an issue about this?

**Comment by mrjbq7** [comment]
Probably worth tracking down that issue also!

**Comment by phred** [comment]
@AlexIljin yeah that sounds worth investigating as a separate issue with the GDI+ backend.

**Comment by phred** [comment]
Looks like there are a couple ways to approach this in CoreGraphics. Whereas `save-ns-image` uses the CoreGraphics function that writes the contents to a file, `image>steam` (called only by `save-graphics-image`) needs to write to the current `output-stream`. The two options are:

1. Write the data into an in-memory buffer with [CGImageDestinationCreateWithData](https://developer.apple.com/documentation/imageio/1465133-cgimagedestinationcreatewithdata), then copy that buffer into `output-stream`
2. Implement `CGDataConsumer` callbacks and use [CGImageDestinationCreateWithDataConsumer](https://developer.apple.com/documentation/imageio/1465231-cgimagedestinationcreatewithdata) to write each chunk directly to `output-stream`.

Of the two I think I like the `CGDataConsumer` approach better, though it requires writing some new CoreGraphics definitions. Seems like `alien-callback` should allow the CoreGraphics code to call into my Factor words, though I admit this isn't something I've worked with very much.
